File: fetchEmails.ts
import { gmail } from './auth';

export interface ParsedEmail {
  id: string;
  from: string;
  subject: string;
  snippet: string;
  timestamp: number;
}

export async function fetchUnreadEmails(limit = 50): Promise<ParsedEmail[]> {
  const res = await gmail.users.messages.list({
    userId: 'me',
    q: '',
    maxResults: limit,
  });

  const messages = res.data.messages || [];

  const parsedEmails: ParsedEmail[] = [];

  for (const msg of messages) {
    const detail = await gmail.users.messages.get({
      userId: 'me',
      id: msg.id!,
    });

    const headers = detail.data.payload?.headers || [];

    const from = headers.find((h: any) => h.name === 'From')?.value || 'Unknown';
    const subject = headers.find((h: any) => h.name === 'Subject')?.value || '(No Subject)';
    const snippet = detail.data.snippet || '';
    const timestamp = Number(detail.data.internalDate) || Date.now();

    parsedEmails.push({
      id: detail.data.id!,
      from,
      subject,
      snippet,
      timestamp,
    });
  }

  return parsedEmails;
}

File: app/fetch_emails.py
from typing import List, Dict, Any
from app.auth import get_gmail_service

def fetch_unread_emails(limit: int = 50) -> List[Dict[str, Any]]:
    """
    Fetch unread emails from Gmail and parse sender, subject, snippet, and timestamp.
    """
    service = get_gmail_service()
    res = service.users().messages().list(userId='me', q='', maxResults=limit).execute()
    messages = res.get('messages', [])
    parsed_emails = []
    for msg in messages:
        detail = service.users().messages().get(userId='me', id=msg['id']).execute()
        headers = detail['payload'].get('headers', [])
        from_ = next((h['value'] for h in headers if h['name'] == 'From'), 'Unknown')
        subject = next((h['value'] for h in headers if h['name'] == 'Subject'), '(No Subject)')
        snippet = detail.get('snippet', '')
        timestamp = int(detail.get('internalDate', '0'))
        parsed_emails.append({
            'id': detail['id'],
            'from': from_,
            'subject': subject,
            'snippet': snippet,
            'timestamp': timestamp
        })
    return parsed_emails
